### ISSUE TITLE
Vault OIDC Setup Issues

### ISSUE THREAD

**Issue by sing011** [hashicorp/vault#14534]
**Describe the bug**
Vault OIDC CLI and GUI not working as expected.

**To Reproduce**
1. vault write auth/oidc/role/app-dev \
   user_claim="email" \
   allowed_redirect_uris="http://localhost:8200/oidc/callback" \
   allowed_redirect_uris="https://vault-dev.smecloudops.com/ui/vault/auth/oidc/oidc/callback"  \
   groups_claim="groups" \
   oidc_scopes="https://graph.microsoft.com/.default" \
   policies=default,adazure

2. vault login -method=oidc role="app-dev"

3. See error

**Expected behavior**
After running vault login -method=oidc role="app-dev", there is a URL on bash. Once I copy paste and go to that URL on browser it asks for user credentials. After putting in the Azure user credentials the callback happens to a local host url on 8250. It should be calling back to the Vault DNS if I'm not wrong. It times out obviously because the URL is localhost and I'm accessing it on internet. 
URL I get - http://localhost:8250/oidc/callback?code=xxxxxxx

The Vault GUI OIDC login does not work either. After selecting that method I hit Login button. Nothing happens when I click it. I tried to curl  http://127.0.0.1:8200/v1/sys/internal/ui/resultant-acl in vault server but got missing client token error.

Also, do I need to change my api_addr to https://vault-dev.smecloudops.com ? Or it shouldn't matter?

**Environment:**
* Vault Server Version (retrieve with `vault status`): 1.7.3
* Vault CLI Version (retrieve with `vault version`): 1.7.3
* Server Operating System/Architecture: Linux

Vault server configuration file(s):

```hcl
ui = true
storage "file" {
  path = "/opt/vault/data"
}

# HTTPS listener
listener "tcp" {
  address = "0.0.0.0:8200"
  cluster_address = "10.12.246.116:8201"
  tls_disable = 1
}

api_addr = "https://vault-prod.smecloudops.com"
cluster_addr = "http://10.12.246.116:8201"

max_lease_ttl = "2160h"
default_lease_ttl = "2160h"
```

**Additional context**

**Comment by austingebauer** [comment]
Hi, @sing011. Happy to help you get this figured out. I think this is likely a configuration issue. Let's focus on the CLI login to start with and then move on to the UI-driven login.

> 3. See error

Are you able to provide the specific error message that you're seeing when logging in via the Vault CLI?

> After putting in the Azure user credentials the callback happens to a local host url on 8250. It should be calling back to the Vault DNS if I'm not wrong. It times out obviously because the URL is localhost and I'm accessing it on internet.

It's expected behavior that the Vault CLI handle the redirect from the browser after you've successfully authenticated to Azure. After the Vault CLI handles the redirect, it will send the request to your Vault server instance to complete the authentication. You'll want to make sure that the `VAULT_ADDR` environment variable is correctly set so that the Vault CLI can complete the login after it handles the browser redirect. Can you check to make sure that's being set appropriately?

**Comment by sing011** [comment]
@austingebauer  Absolutely. So there is no error when I run the cli. I just get the login URL that I put in browser and enter my creds. 

Current VAULT_ADDR = http://127.0.0.1:8200

I tried with http://<server IP>:8200 and also https://<load balancer dns>

**Comment by austingebauer** [comment]
Thanks for the details! If there is no error after running `vault login -method="oidc" role="app-dev"`, then what is the behavior you're seeing? Can you elaborate more on what the issue is?

**Comment by sing011** [comment]
The issue is once I go to the URL on browser and enter my credentials in Azure AD, the next redirect goes to http://localhost:8250/oidc/callback?code=long.random.string.here , since I'm accessing this on internet (browser), the localhost URL just times out as expected. Shouldn't it redirect to the server DNS url(a load balancer endpoint mapped to the server on 8200 ie. https://vault-dev.smecloudops.com/ui/vault/auth/oidc/oidc/callback )?

**Comment by calvn** [comment]
Can you remove `http://localhost:8250/oidc/callback` on the Azure side?

**Comment by sing011** [comment]
@calvn I only had http://localhost:8200/oidc/callback before but had same issue. You want me to remove both and test?

**Comment by sing011** [comment]
@calvn Any update on this? Would appreciate any help

**Comment by artem1982** [comment]
You have typo in config:
allowed_redirect_uris="https://vault-dev.smecloudops.com/ui/vault/auth/oidc/oidc/callback"
allowed_redirect_uris="http://localhost:8200/oidc/callback"

Should be :

allowed_redirect_uris="https://vault-dev.smecloudops.com:8200/ui/vault/auth/oidc/oidc/callback"
allowed_redirect_uris="http://localhost:8250/oidc/callback"


I recommend to switch localhost as number 2 as well for Azure:

allowed_redirect_uris="https://vault-dev.smecloudops.com/ui/vault/auth/oidc/oidc/callback"
allowed_redirect_uris="http://localhost:8250/oidc/callback"

**Comment by austingebauer** [comment]
Were you able to get this sorted out @sing011?
